File: src/api/routes/call_prep.py
"""
Call Prep Pack API Routes

Everything needed to quote and close a lead:
- Property location
- Offer range with justification
- Call script with live injection
"""
from __future__ import annotations

from typing import Any, Dict, Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from api.deps import get_db, get_readonly_db
from core.logging_config import get_logger
from core.models import Lead
from core.address_utils import compute_display_location, compute_mailing_address, format_lead_location_summary
from src.services.offer_helper import compute_offer_range, DEFAULT_DISCOUNT_LOW, DEFAULT_DISCOUNT_HIGH
from src.services.call_script import generate_call_script

router = APIRouter()
LOGGER = get_logger(__name__)


class OfferParamsRequest(BaseModel):
    """Custom offer parameters."""
    discount_low: float = Field(DEFAULT_DISCOUNT_LOW, ge=0.1, le=0.95)
    discount_high: float = Field(DEFAULT_DISCOUNT_HIGH, ge=0.15, le=1.0)


@router.get("/{lead_id}/location")
async def get_lead_location(
    lead_id: int,
    db: Session = Depends(get_readonly_db),
) -> Dict[str, Any]:
    """
    Get normalized location for a lead.
    
    Returns:
    - property_location: Where the land is (situs address or fallback)
    - mailing_address: Where to send mail (owner's address)
    
    These are SEPARATE and should never be confused.
    """
    lead = db.query(Lead).filter(Lead.id == lead_id).first()
    if not lead:
        raise HTTPException(status_code=404, detail="Lead not found")
    
    return format_lead_location_summary(lead)


@router.get("/{lead_id}/offer")
async def get_offer_range(
    lead_id: int,
    discount_low: float = Query(DEFAULT_DISCOUNT_LOW, ge=0.1, le=0.95),
    discount_high: float = Query(DEFAULT_DISCOUNT_HIGH, ge=0.15, le=1.0),
    db: Session = Depends(get_readonly_db),
) -> Dict[str, Any]:
    """
    Get computed offer range for a lead.
    
    NOT comps-dependent. Uses:
    - Land assessed value as basis
    - Acreage for per-acre calculations
    - Condition flags for adjustments
    
    Returns a RANGE with justification bullets (never a single number).
    """
    lead = db.query(Lead).filter(Lead.id == lead_id).first()
    if not lead:
        raise HTTPException(status_code=404, detail="Lead not found")
    
    offer = compute_offer_range(lead, discount_low, discount_high)
    return offer.to_dict()


@router.get("/{lead_id}/script")
async def get_call_script(
    lead_id: int,
    discount_low: float = Query(DEFAULT_DISCOUNT_LOW, ge=0.1, le=0.95),
    discount_high: float = Query(DEFAULT_DISCOUNT_HIGH, ge=0.15, le=1.0),
    db: Session = Depends(get_readonly_db),
) -> Dict[str, Any]:
    """
    Generate call script for a lead.
    
    All values are injected from lead data:
    - Property location
    - Acreage
    - Offer range
    
    Script updates live when discount params change.
    """
    lead = db.query(Lead).filter(Lead.id == lead_id).first()
    if not lead:
        raise HTTPException(status_code=404, detail="Lead not found")
    
    script = generate_call_script(lead, discount_low, discount_high)
    return script.to_dict()


@router.get("/{lead_id}/prep-pack")
async def get_call_prep_pack(
    lead_id: int,
    discount_low: float = Query(DEFAULT_DISCOUNT_LOW, ge=0.1, le=0.95),
    discount_high: float = Query(DEFAULT_DISCOUNT_HIGH, ge=0.15, le=1.0),
    db: Session = Depends(get_readonly_db),
) -> Dict[str, Any]:
    """
    Get complete Call Prep Pack for a lead.
    
    Everything needed to quote and close in one response:
    - Location (property + mailing, clearly separated)
    - Parcel snapshot
    - Offer range with justification
    - Call script with live injection
    - Map data
    """
    lead = db.query(Lead).filter(Lead.id == lead_id).first()
    if not lead:
        raise HTTPException(status_code=404, detail="Lead not found")
    
    # Get all components
    location = format_lead_location_summary(lead)
    offer = compute_offer_range(lead, discount_low, discount_high)
    script = generate_call_script(lead, discount_low, discount_high)
    
    # Parcel snapshot
    parcel = lead.parcel
    parcel_snapshot = {
        "parcel_id": parcel.canonical_parcel_id if parcel else None,
        "parish": parcel.parish if parcel else None,
        "acreage": float(parcel.lot_size_acres) if parcel and parcel.lot_size_acres else None,
        "land_value": float(parcel.land_assessed_value) if parcel and parcel.land_assessed_value else None,
        "is_adjudicated": parcel.is_adjudicated if parcel else False,
        "years_tax_delinquent": parcel.years_tax_delinquent if parcel else 0,
    }
    
    # Map data
    map_data = {
        "has_coordinates": bool(parcel and parcel.latitude and parcel.longitude),
        "latitude": parcel.latitude if parcel else None,
        "longitude": parcel.longitude if parcel else None,
        "geocode_needed": parcel and not parcel.latitude and not parcel.longitude,
    }
    
    # Owner info
    owner = lead.owner
    party = owner.party if owner else None
    owner_info = {
        "name": party.display_name if party else "Unknown",
        "phone": owner.phone_primary if owner else None,
        "email": owner.email if owner else None,
        "is_tcpa_safe": owner.is_tcpa_safe if owner else False,
    }
    
    return {
        "lead_id": lead.id,
        "motivation_score": lead.motivation_score,
        "pipeline_stage": lead.pipeline_stage,
        "owner": owner_info,
        "location": location,
        "parcel": parcel_snapshot,
        "offer": offer.to_dict(),
        "script": script.to_dict(),
        "map": map_data,
    }


__all__ = ["router"]

File: frontend/src/api/callPrep.ts
/**
 * Call Prep Pack API - PRODUCTION HARDENED
 * 
 * Everything needed to quote and close a lead:
 * - Location (property + mailing, clearly separated with trust indicators)
 * - Offer range with justification and explicit warnings
 * - Call script with live injection
 * 
 * CRITICAL: Property location and mailing address are SEPARATE.
 * Never confuse them.
 */
import { apiClient } from './client';

// Data Trust Levels
export type DataTrust = 'verified_gis' | 'parcel_record' | 'derived' | 'owner_provided' | 'missing';

// Data Warnings
export type DataWarning = 
  | 'no_situs_address' 
  | 'no_coordinates' 
  | 'no_parcel' 
  | 'no_geometry' 
  | 'geocode_needed' 
  | 'mailing_only' 
  | 'unverified_location';

// Offer Warnings
export type OfferWarning = 
  | 'missing_land_value'
  | 'missing_acreage'
  | 'zero_land_value'
  | 'adjudicated_title_risk'
  | 'tax_delinquent'
  | 'no_parcel_data'
  | 'estimate_only';

// Types
export interface PropertyLocation {
  address_line1: string | null;
  city: string | null;
  state: string | null;
  postal_code: string | null;
  parish: string;
  parcel_id: string;
  latitude: number | null;
  longitude: number | null;
  data_trust: DataTrust;
  has_situs_address: boolean;
  has_coordinates: boolean;
  has_geometry: boolean;
  full_address: string;
  short_address: string;
  location_descriptor: string;
  map_query: string | null;
  assessor_search_query: string;
  can_show_map: boolean;
  map_trust_level: string;
  missing_data_message: string | null;
  warnings: DataWarning[];
}

export interface MailingAddress {
  line1: string | null;
  line2: string | null;
  city: string | null;
  state: string | null;
  postal_code: string | null;
  raw_address: string | null;
  display: string;
  is_available: boolean;
  data_trust: DataTrust;
  _warning?: string;
}

export interface LocationSummary {
  property_location: PropertyLocation;
  mailing_address: MailingAddress;
  _trust_warning?: string;
}

export interface JustificationBullet {
  factor: string;
  description: string;
  impact: 'increase' | 'decrease' | 'neutral';
}

export type OfferConfidence = 'high' | 'medium' | 'low' | 'cannot_compute';

export interface OfferRange {
  low_offer: number;
  high_offer: number;
  midpoint: number;
  range_display: string;
  land_value: number | null;
  acreage: number | null;
  discount_low: number;
  discount_high: number;
  price_per_acre_low: number | null;
  price_per_acre_high: number | null;
  per_acre_display: string | null;
  justifications: JustificationBullet[];
  confidence: OfferConfidence;
  confidence_reason: string;
  warnings: OfferWarning[];
  missing_data_summary: string | null;
  can_make_offer: boolean;
  cannot_offer_reason: string | null;
}

export interface ObjectionHandler {
  objection: string;
  response: string;
}

export interface CallScript {
  lead_id: number;
  owner_name: string;
  property_location: string;
  acreage: number | null;
  parish: string;
  offer_range: OfferRange;
  opening: string;
  discovery: string;
  price_discussion: string;
  objection_handlers: ObjectionHandler[];
  closing: string;
}

export interface ParcelSnapshot {
  parcel_id: string | null;
  parish: string | null;
  acreage: number | null;
  land_value: number | null;
  is_adjudicated: boolean;
  years_tax_delinquent: number;
}

export interface MapData {
  has_coordinates: boolean;
  latitude: number | null;
  longitude: number | null;
  geocode_needed: boolean;
}

export interface OwnerInfo {
  name: string;
  phone: string | null;
  email: string | null;
  is_tcpa_safe: boolean;
}

export interface CallPrepPack {
  lead_id: number;
  motivation_score: number;
  pipeline_stage: string;
  owner: OwnerInfo;
  location: LocationSummary;
  parcel: ParcelSnapshot;
  offer: OfferRange;
  script: CallScript;
  map: MapData;
}

/**
 * Get normalized location for a lead
 */
export async function getLeadLocation(leadId: number): Promise<LocationSummary> {
  const response = await apiClient.get<LocationSummary>(`/call-prep/${leadId}/location`);
  return response.data;
}

/**
 * Get computed offer range for a lead
 */
export async function getOfferRange(
  leadId: number,
  discountLow?: number,
  discountHigh?: number
): Promise<OfferRange> {
  const params: Record<string, number> = {};
  if (discountLow !== undefined) params.discount_low = discountLow;
  if (discountHigh !== undefined) params.discount_high = discountHigh;
  
  const response = await apiClient.get<OfferRange>(`/call-prep/${leadId}/offer`, { params });
  return response.data;
}

/**
 * Get call script for a lead
 */
export async function getCallScript(
  leadId: number,
  discountLow?: number,
  discountHigh?: number
): Promise<CallScript> {
  const params: Record<string, number> = {};
  if (discountLow !== undefined) params.discount_low = discountLow;
  if (discountHigh !== undefined) params.discount_high = discountHigh;
  
  const response = await apiClient.get<CallScript>(`/call-prep/${leadId}/script`, { params });
  return response.data;
}

/**
 * Get complete Call Prep Pack for a lead
 */
export async function getCallPrepPack(
  leadId: number,
  discountLow?: number,
  discountHigh?: number
): Promise<CallPrepPack> {
  const params: Record<string, number> = {};
  if (discountLow !== undefined) params.discount_low = discountLow;
  if (discountHigh !== undefined) params.discount_high = discountHigh;
  
  const response = await apiClient.get<CallPrepPack>(`/call-prep/${leadId}/prep-pack`, { params });
  return response.data;
}
